sys/kern/src/arch/riscv64.rs

```diff
@@ -0,0 +1,603 @@
+                task::NextTask::Other => task::select(idx, tasks),
+                task::NextTask::Same => idx,
+            };
+
+            if next == idx {
+                panic!("attempt to return to Task #{} after fault", idx);
+            }
+
+            let next = &mut tasks[next];
+            apply_memory_protection(next);
+            // Safety: this leaks a pointer aliasing into static scope, but
+            // we're not going to read it back until the next kernel entry so
+            // we won't be aliasing/racing.
+            set_current_task(next);
+        });
+    }
+}
+
+// Timer handling.
+//
+// We currently only support single HART systems.  From reading elsewhere,
+// additional harts have their own mtimecmp offset at 0x8 intervals from hart0.
+//
+// As per FE310-G002 Manual, section 9.1, the address of mtimecmp on
+// our supported board is 0x0200_4000, which also matches qemu.
+//
+// On both RV32 and RV64 systems the mtime and mtimecmp memory-mapped registers
+// are 64-bits wide.
+//
+const MTIMECMP: u64 = 0x0200_4000;
+const MTIME: u64 = 0x0200_BFF8;
+
+// Configure the timer.
+//
+// RISC-V Privileged Architecture Manual
+// 3.2.1 Machine Timer Registers (mtime and mtimecmp)
+//
+// To keep things simple, especially on RV32 systems where we cannot atomically
+// write to the mtime/mtimecmp memory-mapped registers as they are 64 bits
+// wide, we only utilise the first 32-bits of each register, setting the
+// high-order bits to 0 on startup, and restarting the low-order bits of mtime
+// back to 0 on each interrupt.
+//
+#[no_mangle]
+unsafe fn set_timer(tick_divisor: u32) {
+    // Set high-order bits of mtime to zero.  We only call this function prior
+    // to enabling interrupts so it should be safe.
+    unsafe {
+        asm!("
+        li {0}, {mtimecmp}  # load mtimecmp address
+        sd {1}, 0({0})      # set mtimecmp register
+
+        li {0}, {mtime}     # load mtime address
+        sd zero, 0({0})     # set low-order bits back to 0
+        ",
+            out(reg) _,
+            in(reg) tick_divisor,
+            mtime = const MTIME,
+            mtimecmp = const MTIMECMP,
+        );
+    }
+}
+
+#[allow(unused_variables)]
+pub fn start_first_task(tick_divisor: u32, task: &task::Task) -> ! {
+    // Configure MPP to switch us to User mode on exit from Machine
+    // mode (when we call "mret" below).
+    unsafe {
+        register::mstatus::set_mpp(MPP::User);
+    }
+
+    // Write the initial task program counter.
+    register::mepc::write(task.save().pc as *const usize as usize);
+
+    //
+    // Configure the timer
+    //
+    unsafe {
+        CLOCK_FREQ_KHZ = tick_divisor;
+
+        // Reset mtime back to 0, set mtimecmp to chosen timer
+        set_timer(tick_divisor - 1);
+
+        // Machine timer interrupt enable
+        register::mie::set_mtimer();
+
+        // Global machine interrupt enable
+        register::mstatus::set_mie();
```

Comment:
Setting `MPIE` would set previous-mode's interrupt enable bit on mret. However, at this point, the aim is to enable interrupts for machine mode. I don't think I understood what bug you are referring to. Is the bug because `CURRENT_TASK_PTR` is not set? If so, then `CURRENT_TASK_PTR` can be set before enabling machine interrupts. But, the intent here was to enable machine interrupts.

Reply:
I think I see what you are saying. We don't really want machine-mode interrupts to be enabled while we are still in M-mode. We are basically running without interrupts enabled in M-mode all the time. Is there any benefit of setting MPIE at all for us? We don't really support nested interrupts and interrupts to M-mode are default enabled when going to lower modes (S/U-mode).

Reply:
So we don't actually want to enable interrupts yet.  On line 525 we set `mepc` so it is ready when we `mret` to the first task.  However, any interrupt will change mepc.  If we instead just set `mpie`, after we `mret`, all interrupts will be enabled.  
Aside from this, I think interrupts are always disabled when in m-mode for hubris.  The only way to get into m-mode is through the trap handler, which disables on entry, and re enables on exit.

Reply:
Done. Pushed a change to fix this: https://github.com/rivosinc/hubris/pull/17/commits/0550dbb00f9d70bbbc45cb63b189461f4d267601

Reply:
Looks like I missed your recent response. 

>  If we instead just set mpie, after we mret, all interrupts will be enabled.

We don't really need to set MPIE as well. Interrupts to M-mode are default enabled in all lower modes.

Reply:
I saw you latest push.  I forget we are not delegating interrupts so MPIE doesn't really matter here.  I didn't realize m-mode interrupts were always enables when in a lower privilege mode either.  Will make a patch up for the main rv32 port to correct this.